pkg/proxy/iptables/proxier.go

```diff
@@ -572,45 +590,56 @@ func (proxier *Proxier) OnEndpointsSynced() {
 	proxier.syncProxyRules(syncReasonEndpoints)
 }
 
-// Convert a slice of api.Endpoints objects into a map of service-port -> endpoints.
-func buildNewEndpointsMap(allEndpoints endpointsMap, curMap proxyEndpointMap, hostname string) (newMap proxyEndpointMap, hcEndpoints map[types.NamespacedName]int, staleSet map[endpointServicePair]bool) {
+func (proxier *Proxier) mergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		proxier.endpointsMap[svcPort] = endpointsMap[svcPort]
+	}
 
-	// return values
-	newMap = make(proxyEndpointMap)
-	hcEndpoints = make(map[types.NamespacedName]int)
-	staleSet = make(map[endpointServicePair]bool)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn, ips := range localIPs {
+		proxier.hcEndpoints[nsn] = len(ips)
+	}
+}
 
-	// Update endpoints for services.
-	for _, endpoints := range allEndpoints {
-		accumulateEndpointsMap(endpoints, hostname, &newMap)
+func (proxier *Proxier) unmergeEndpointsMap(endpointsMap proxyEndpointsMap) {
+	for svcPort := range endpointsMap {
+		delete(proxier.endpointsMap, svcPort)
 	}
-	// Check stale connections against endpoints missing from the update.
-	// TODO: we should really only mark a connection stale if the proto was UDP
-	// and the (ip, port, proto) was removed from the endpoints.
-	for svcPort, epList := range curMap {
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ExternalTrafficLocalOnly) {
+		return
+	}
+	localIPs := getLocalIPs(endpointsMap)
+	for nsn := range localIPs {
+		delete(proxier.hcEndpoints, nsn)
+	}
+}
+
+func (proxier *Proxier) detectStaleConnections(oldEndpointsMap, newEndpointsMap proxyEndpointsMap) {
```

Comment:
Actually, we will need to store both "version from before the first change (if there were multiple)" and the currently desired version. But that doesn't change much, and is still doable.

I agree that it sounds like a better option.

Will change next week.